## **Honour Code**
I Siphesihle, Yapi, confirm - by submitting my - that the solutions in this notebook are a result of my own work and that I abide by the EDSA honour code (https://drive.google.com/file/d/1QDCjGZJ8-FmJE3bZdIQNwnJyQKPhHZBn/view?usp=sharing).

Non-compliance with the honour code constitutes a material breach of contract.

In [0]:
import numpy as np
import pandas as pd
from sklearn.metrics import classification_report
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier

In [0]:
train  = pd.read_csv('/content/drive/My Drive/zindi/Xente/training.csv')
test = pd.read_csv('/content/drive/My Drive/zindi/Xente/test.csv')

In [44]:
print(train.shape)
train.head()

(95662, 16)


TransactionId         BatchId  ... PricingStrategy FraudResult
0  TransactionId_76871   BatchId_36123  ...               2           0
1  TransactionId_73770   BatchId_15642  ...               2           0
2  TransactionId_26203   BatchId_53941  ...               2           0
3    TransactionId_380  BatchId_102363  ...               2           0
4  TransactionId_28195   BatchId_38780  ...               2           0

[5 rows x 16 columns]

In [45]:
print(test.shape)
test.head()

(45019, 15)


TransactionId        BatchId  ...  TransactionStartTime PricingStrategy
0  TransactionId_50600  BatchId_35028  ...  2019-02-13T10:01:40Z               4
1  TransactionId_95109  BatchId_45139  ...  2019-02-13T10:02:12Z               2
2  TransactionId_47357  BatchId_74887  ...  2019-02-13T10:02:30Z               2
3  TransactionId_28185  BatchId_11025  ...  2019-02-13T10:02:38Z               4
4  TransactionId_22140  BatchId_29804  ...  2019-02-13T10:02:58Z               2

[5 rows x 15 columns]

In [0]:
y = train['FraudResult']

In [0]:
new_train = train.drop('FraudResult',axis =1)

In [0]:
combined = new_train.append(test,sort = False)

In [0]:
len_train = len(train)
len_test = len(test)

In [50]:
combined['TransactionId'].replace(regex=True,inplace=True,to_replace=r'\D',value=r'')
combined['BatchId'].replace(regex=True,inplace=True,to_replace=r'\D',value=r'')
combined['AccountId'].replace(regex=True,inplace=True,to_replace=r'\D',value=r'')
combined['SubscriptionId'].replace(regex=True,inplace=True,to_replace=r'\D',value=r'')
combined['CustomerId'].replace(regex=True,inplace=True,to_replace=r'\D',value=r'')
combined['ProviderId'].replace(regex=True,inplace=True,to_replace=r'\D',value=r'')
combined['ProductId'].replace(regex=True,inplace=True,to_replace=r'\D',value=r'')
combined['ChannelId'].replace(regex=True,inplace=True,to_replace=r'\D',value=r'')
combined = combined.drop(['CountryCode','CurrencyCode','ProductCategory','TransactionStartTime'],axis =1)
combined.head()

TransactionId BatchId AccountId  ...   Amount  Value PricingStrategy
0         76871   36123      3957  ...   1000.0   1000               2
1         73770   15642      4841  ...    -20.0     20               2
2         26203   53941      4229  ...    500.0    500               2
3           380  102363       648  ...  20000.0  21800               2
4         28195   38780      4841  ...   -644.0    644               2

[5 rows x 11 columns]

In [0]:
X = combined[:len_train]
real_X = combined[len_train:]

In [0]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [53]:
lm = LogisticRegression()
lm.fit(X_train,y_train)
pred_lm = lm.predict(X_test)
accuracy_score(y_test,pred_lm)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


0.9982752312758062

In [54]:
print(classification_report(y_test, pred_lm))


              precision    recall  f1-score   support

           0       1.00      1.00      1.00     19097
           1       0.62      0.22      0.33        36

    accuracy                           1.00     19133
   macro avg       0.81      0.61      0.66     19133
weighted avg       1.00      1.00      1.00     19133



In [55]:
ran = RandomForestClassifier()
ran.fit(X_train,y_train)
y_ran = ran.predict(X_test)
accuracy_score(y_test,y_ran)

/usr/local/lib/python3.6/dist-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


0.9997909371243402

In [56]:
print(classification_report(y_test, y_ran))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     19097
           1       1.00      0.89      0.94        36

    accuracy                           1.00     19133
   macro avg       1.00      0.94      0.97     19133
weighted avg       1.00      1.00      1.00     19133



In [57]:
knn = KNeighborsClassifier()
knn.fit(X_train,y_train)
y_knn = knn.predict(X_test)
accuracy_score(y_test,y_knn)

0.9996341399675952

In [58]:
print(classification_report(y_test, y_knn))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     19097
           1       0.94      0.86      0.90        36

    accuracy                           1.00     19133
   macro avg       0.97      0.93      0.95     19133
weighted avg       1.00      1.00      1.00     19133



In [0]:
test['FraudResult_ran'] = ran.predict(real_X)

In [62]:
submission = test[['TransactionId','FraudResult_ran']]
submission.head()

TransactionId  FraudResult_ran
0  TransactionId_50600                0
1  TransactionId_95109                0
2  TransactionId_47357                0
3  TransactionId_28185                0
4  TransactionId_22140                0

In [0]:
from google.colab import files
exportTest = submission
exportTest.to_csv('submission.csv',index=False) 
files.download('submission.csv')

**Conclusion**

RandomForestClassifier gave the better accuracy and a better classification report. With it i got a score of 0.66 on Zindi